gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/adwords/GoogleAdWordsExtractor.java

```diff
@@ -0,0 +1,262 @@
+    typeConversionMap.put("float", JsonElementConversionFactory.Type.FLOAT);
+    typeConversionMap.put("double", JsonElementConversionFactory.Type.DOUBLE);
+    typeConversionMap.put("boolean", JsonElementConversionFactory.Type.BOOLEAN);
+  }
+
+  public GoogleAdWordsExtractor(WorkUnitState state)
+      throws Exception {
+    _state = state;
+    long lowWatermark = state.getWorkunit().getLowWatermark(LongWatermark.class).getValue();
+    _startDate = watermarkFormatter.parseDateTime(Long.toString(lowWatermark));
+    long highWatermark = state.getWorkunit().getExpectedHighWatermark(LongWatermark.class).getValue();
+    _expectedEndDate = watermarkFormatter.parseDateTime(Long.toString(highWatermark));
+
+    GoogleAdWordsCredential credential = new GoogleAdWordsCredential(state);
+    AdWordsSession.ImmutableAdWordsSession rootSession = credential.buildRootSession();
+
+    ReportDefinitionReportType reportType =
+        ReportDefinitionReportType.valueOf(state.getProp(GoogleAdWordsSource.KEY_REPORT).toUpperCase() + "_REPORT");
+    ReportDefinitionDateRangeType dateRangeType =
+        ReportDefinitionDateRangeType.valueOf(state.getProp(GoogleAdWordsSource.KEY_DATE_RANGE).toUpperCase());
+    if (!dateRangeType.equals(ReportDefinitionDateRangeType.CUSTOM_DATE) && !dateRangeType
+        .equals(ReportDefinitionDateRangeType.ALL_TIME)) {
+      throw new UnsupportedOperationException("Only support date range of custom_date or all_time");
+    }
+
+    String columnNamesString = state.getProp(GoogleAdWordsSource.KEY_COLUMN_NAMES, "");
+    List<String> columnNames =
+        columnNamesString.trim().isEmpty() ? null : Lists.newArrayList(splitter.split(columnNamesString));
+
+    HashMap<String, String> allFields = downloadReportFields(rootSession, reportType);
+    try {
+      _schema = createSchema(allFields, columnNames);
+      log.info(String.format("Schema for report %s: %s", reportType, _schema));
+    } catch (IOException e) {
+      throw new RuntimeException(String.format("Failed downloading report %s", reportType), e);
+    }
+
+    _iterator = new GoogleAdWordsExtractorIterator(
+        new GoogleAdWordsReportDownloader(rootSession, _state, dateFormatter.print(_startDate),
+            dateFormatter.print(_expectedEndDate), reportType, dateRangeType, _schema),
+        getConfiguredAccounts(rootSession, state));
+  }
+
+  /**
+   * 1. Get all available non-manager accounts.
+   * 2. If exactAccounts are provided, validate that all exactAccounts are a subset of step 1.
+   * 3. If exclusiveAccounts are provided, remove them from step 1.
+   */
+  static Collection<String> getConfiguredAccounts(AdWordsSession rootSession, WorkUnitState state)
+      throws ValidationException, RemoteException {
+
+    String masterCustomerId = state.getProp(GoogleAdWordsSource.KEY_MASTER_CUSTOMER);
+    String exactAccountString = state.getProp(GoogleAdWordsSource.KEY_ACCOUNTS_EXACT, "");
+    Set<String> exactAccounts =
```

Comment:
nit: change to: 

Optional<String> optionalExactAccountsString = Optional.forNullable(state.getProp(GoogleAdWordsSource.KEY_ACCOUNTS_EXACT, null)); 

Optional<Set<String>> optionalExactAccounts = optionalExactAccounts.isPresent() ? Sets.newHashSet(SPLIT_BY_COMMA.split(optionalExactAccounts.get())) : Optional.absent(); 

.. and so on